keyboards/skyloong/gk87/bl_pro/bl_pro.c

```diff
@@ -0,0 +1,192 @@
+
+       case KC_LGUI:
+           if (FN_ON){
+              if ( record->event.pressed){
+                 WIN_LOCK = !WIN_LOCK ; //change win lock state
+                }
+              if (!WIN_LOCK) {
+                 IND = IND & (~WINLK_ON);   //Close win lock display
+                 s_serial_to_parallel(IND);
+                 gpio_write_pin_high(WIN_LOCK_PIN);
+                 return false; //windows key locked do nothing
+                }
+              s_serial_to_parallel(IND);
+            }
+           if (WIN_LOCK) {
+               IND = IND | WINLK_ON;   //Open win lock display
+               s_serial_to_parallel(IND);
+               gpio_write_pin_low(WIN_LOCK_PIN);
+               return false; //windows key locked do nothing
+            }
+           return true;  // continue all further processing of this key
+
+       case DF(0):
+           set_single_persistent_default_layer(0);
+           return true;
+
+       case DF(1):
+           set_single_persistent_default_layer(1);
+           return true;
+
+       case BL_TOGG:
+           if (record->event.pressed){
+               if(is_backlight_breathing()) {
+                    backlight_disable_breathing();
+                    DIS_BRETH = 1;
+
+                }else if(DIS_BRETH && !(is_backlight_enabled())){
+                    backlight_enable_breathing();
+                    DIS_BRETH = 0;
+                }
+           }
+         return true;
+
+       case BL_BRTG:
+           if (record->event.pressed){
+               if(DIS_BRETH || !(is_backlight_enabled())) {
+                 return false;
+                }
+           }
+         return true;
+
+       case BL_UP:
+           DIS_BRETH = 0;
+         return true;
+
+       case BL_DOWN:
+           if (record->event.pressed){
+               if(DIS_BRETH || !(is_backlight_enabled())) {
+                 return false;
+                }
+           }
+         return true;
+
+    default:
+      return true;
+    }
+   s_serial_to_parallel(IND);
+}
+
+void suspend_power_down_kb() {
+    SLEEP = 1;
+    suspend_power_down_user();
+}
+
+void suspend_wakeup_init_kb() {
+    SLEEP = 0;
+    gpio_write_pin(LED_CAPS_LOCK_PIN, !host_keyboard_led_state().caps_lock);
+    gpio_write_pin(LED_SCROLL_LOCK_PIN, !host_keyboard_led_state().scroll_lock);
+    gpio_write_pin(WIN_LOCK_PIN, !WIN_LOCK);
+    s_serial_to_parallel(IND);
+    suspend_wakeup_init_user();
+}
+
+bool shutdown_kb(bool jump_to_bootloader) {
+    SLEEP = 1;
+    return true;
+}
+
+layer_state_t default_layer_state_set_kb(layer_state_t state) {
+    switch (get_highest_layer(state)) {
+    case 0:
+        //switch to win layer display
+        IND = IND & (~MAC_ON);
+        IND = IND | WIN_ON;
+        break;
+    case 1:
+        //switch to mac layer display
+        IND = IND & (~WIN_ON);
+        IND = IND | MAC_ON;
+        break;
+    }
+    s_serial_to_parallel(IND);
+  return default_layer_state_set_user(state);
```

Comment:
It may be better to handle the user callback first, that way, the indicator bitmask reflects any changes set in user code. 
```suggestion
layer_state_t default_layer_state_set_kb(layer_state_t state) {
    state = default_layer_state_set_user(state);
    switch (get_highest_layer(state)) {
    case 0:
        //switch to win layer display
        IND = IND & (~MAC_ON);
        IND = IND | WIN_ON;
        break;
    case 1:
        //switch to mac layer display
        IND = IND & (~WIN_ON);
        IND = IND | MAC_ON;
        break;
    }
    s_serial_to_parallel(IND);
    return state;
```